### PR TITLE
Use named fields for `{ast,hir}::ItemKind::Impl`

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Currently, the widely used `ItemKind::Impl` variant is a tuple with seven fields. I want to add an eighth in #68140, which means I have to update most matches on this variant anyways. Giving a name to each field improves readability and makes future changes of this nature much simpler.

This change will cause several tools to break. I will fix them once this is merged.

### CODE REVIEW

**Comment by oli-obk** [comment]
cc @petrochenkov @Centril

**Comment by Centril** [comment]
Seems fine to me; just make sure to use exhaustive matching where appropriate by enumerating the fields with `field: _`, but it seems like that's already done.

File: src/libsyntax/visit.rs [hunk 1]
```diff
@@ -308,11 +308,19 @@ pub fn walk_item<'a, V: Visitor<'a>>(visitor: &mut V, item: &'a Item) {
             visitor.visit_generics(generics);
             visitor.visit_enum_def(enum_definition, generics, item.id, item.span)
         }
-        ItemKind::Impl(_, _, _, ref generics, ref opt_trait_reference, ref typ, ref impl_items) => {
+        ItemKind::Impl {
+            unsafety: _,
+            polarity: _,
+            defaultness: _,
+            ref generics,
+            ref of_trait,
```
Comment (Centril) [inline src/libsyntax/visit.rs]: Not sure what "of" signifies here. I would rename this to `trait_ref` and let the type system speak to the `Option<_>`

Comment (ecstatic-morse) [reply src/libsyntax/visit.rs]: It's `Some` if the item is an impl of a trait.

Comment (Centril) [reply src/libsyntax/visit.rs]: But what does the f stand for? (I know what the field means.)

Comment (ecstatic-morse) [reply src/libsyntax/visit.rs]: "of" like the preposition. An impl *of* a trait.

Comment (Centril) [reply src/libsyntax/visit.rs]: Oh. Well that was far from obvious. :) Given that you had to ELI5 it for me, could we change it to something else?

Comment (ecstatic-morse) [reply src/libsyntax/visit.rs]: I could, although to be honest I'm not enthusiastic about it. On a personal note, I have less time to contribute to open source this year, and having to spend it arguing over trivial choices like this has left me feeling a bit bummed out 😢.

Comment (Centril) [reply src/libsyntax/visit.rs]: (Fwiw, I didn't think we were arguing -- I was just confused as to what it signified.)

**Comment by oli-obk** [comment]
In case this bitrots, rebase and r=me with p=1
